lib/backend/repository/repository2.go

```diff
@@ -0,0 +1,370 @@
+	}
+	defer os.RemoveAll(tmpDir)
+
+	layerFile, err := rb.getLayerV2(layerID, dockerURL, tmpDir)
+	if err != nil {
+		return "", nil, fmt.Errorf("error getting the remote layer: %v", err)
+	}
+	defer layerFile.Close()
+
+	util.Debug("Generating layer ACI...")
+	aciPath, aciManifest, err := common.GenerateACI(layerNumber, layerData, dockerURL, outputDir, layerFile, curPwl, compress)
+	if err != nil {
+		return "", nil, fmt.Errorf("error generating ACI: %v", err)
+	}
+
+	return aciPath, aciManifest, nil
+}
+
+func (rb *RepositoryBackend) getManifestV2(dockerURL *types.ParsedDockerURL) (*v2Manifest, []string, error) {
+	url := rb.protocol() + path.Join(dockerURL.IndexURL, "v2", dockerURL.ImageName, "manifests", dockerURL.Tag)
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if rb.username != "" && rb.password != "" {
+		req.SetBasicAuth(rb.username, rb.password)
+	}
+
+	res, err := rb.makeRequest(req, dockerURL.ImageName)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, nil, fmt.Errorf("unexpected http code: %d, URL: %s", res.StatusCode, req.URL)
+	}
+
+	manblob, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	manifest := &v2Manifest{}
+
+	err = json.Unmarshal(manblob, manifest)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if manifest.Name != dockerURL.ImageName {
+		return nil, nil, fmt.Errorf("name doesn't match what was requested, expected: %s, downloaded: %s", dockerURL.ImageName, manifest.Name)
+	}
+
+	if manifest.Tag != dockerURL.Tag {
+		return nil, nil, fmt.Errorf("tag doesn't match what was requested, expected: %s, downloaded: %s", dockerURL.Tag, manifest.Tag)
+	}
+
+	//TODO: verify signature here
+
+	layers := make([]string, len(manifest.FSLayers))
+
+	for i, layer := range manifest.FSLayers {
+		layers[i] = layer.BlobSum
+	}
+
+	return manifest, layers, nil
+}
+
+func getLayerIndex(layerID string, manifest v2Manifest) (int, error) {
+	for i, layer := range manifest.FSLayers {
+		if layer.BlobSum == layerID {
+			return i, nil
+		}
+	}
+	return -1, fmt.Errorf("layer not found in manifest: %s", layerID)
+}
+
+func (rb *RepositoryBackend) getLayerV2(layerID string, dockerURL *types.ParsedDockerURL, tmpDir string) (*os.File, error) {
+	url := rb.protocol() + path.Join(dockerURL.IndexURL, "v2", dockerURL.ImageName, "blobs", layerID)
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := rb.makeRequest(req, dockerURL.ImageName)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode != 200 {
+		return nil, fmt.Errorf("HTTP code: %d. URL: %s", res.StatusCode, req.URL)
+	}
+
+	var in io.Reader
+	in = res.Body
+
+	if hdr := res.Header.Get("Content-Length"); hdr != "" {
+		imgSize, err := strconv.ParseInt(hdr, 10, 64)
+		if err != nil {
+			return nil, err
+		}
+
+		prefix := "Downloading " + layerID[:12]
```

Comment:
We can increase a bit the length of the printed layerID because it includes `sha256:` now.